### PR TITLE
InstCombine: fix transformation of bitwiseAnd to fabs

### LINKED ISSUE #69464: InstCombine generates "fabs" calls from "And" pattern that is not correct in OpenCL
Since llvm/llvm-project@5c0da5839de1 instcombine transform pattern masking the MSB with a call to fabs.

While this might be correct in C, it is not in OpenCL. Math operations are allowed to flush denormalized numbers to zero in OpenCL. Thus what should be just masking the MSB starts to change much more data.

This is an issue in libclc as those implementations are using this pattern and rely on the fact that only the MSB is masked. Using fabs instead breaks conformance on platforms where fabs flushes denormalized numbers to zero.

### PR DESCRIPTION
**Pull request by rjodinchr**
Since 5c0da5839de1 instcombine transform pattern masking the MSB with a call to fabs.

While this might be correct in C, it is not in OpenCL. Math operations are allowed to flush denormalized numbers to zero in OpenCL. Thus what should be just masking the MSB starts to change much more data.

This is an issue in libclc as those implementations are using this pattern and rely on the fact that only the MSB is masked. Using fabs instead breaks conformance on platforms where fabs flushes denormalized numbers to zero.

Instead look for the complete pattern:
```
bitcast float to int
bitwiseand
bitcast int to float
=>
fabs
```

Fix #69464

### CODE REVIEW

**Comment by rjodinchr** [comment]
@arsenm @alan-baker , Could you please review?

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 1]
```diff
@@ -2546,6 +2546,32 @@ Instruction *InstCombinerImpl::visitBitCast(BitCastInst &CI) {
...
+  // floating-point operation.
+  //
+  // Assumes any IEEE-represented type has the sign bit in the high bit.
+  if (!Builder.GetInsertBlock()->getParent()->hasFnAttribute(
+          Attribute::NoImplicitFloat) &&
+      DestTy->isFloatingPointTy() && DestTy->isIEEE()) {
+    Value *L, *R;
+    if (match(Src, m_And(m_Value(L), m_Value(R)))) {
+      Value *Cast;
+      if (match(L, m_BitCast(m_Value(Cast))) && match(R, m_MaxSignedValue())) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:2562]:
```suggestion
    Value *Cast;
    if (match(Src, m_And(m_BitCast(m_Value(Cast)), m_MaxSignedValue()))) {
```

**Review by arsenm — CHANGES_REQUESTED** [review]
The rationale for this change is incorrect. By definition, fabs, fneg, and copysign only change the sign bit and cannot make any other value modifications. Any lowering of these operations which do flush denormals or quiet signaling nans are incorrect and should be fixed. This is universally true, OpenCL is not unique here.

**Comment by rjodinchr** [comment]
Could you please provide references of this universal definition that `fabs` only changed the sign bit?

**Comment by nikic** [comment]
> Could you please provide references of this universal definition that `fabs` only changed the sign bit?

Quoting from LangRef (https://llvm.org/docs/LangRef.html#llvm-fabs-intrinsic):

> This function returns the same values as the libm fabs functions would, and handles error conditions in the same way. The returned value is completely identical to the input except for the sign bit; in particular, if the input is a NaN, then the quiet/signaling bit and payload are perfectly preserved.

**Pull request closed without merging**